### PR TITLE
GH-3950 SHACL validation against data in an existing Sail

### PR DESCRIPTION
**Pull request by hmottestad**
GitHub issue resolved: #3950 

Briefly describe the changes proposed in this PR:

----
PR Author Checklist (see the [contributor guidelines](https://github.com/eclipse/rdf4j/blob/main/.github/CONTRIBUTING.md) for more details):

 - [ ] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [ ] I've added tests for the changes I made
 - [ ] I've applied [code formatting](https://github.com/eclipse/rdf4j/blob/main/.github/CONTRIBUTING.md#code-formatting) (you can use `mvn process-resources` to format from the command line)
 - [ ] I've [squashed](https://rdf4j.org/documentation/developer/squashing) my commits where necessary 
 - [ ] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change

### CODE REVIEW

File: examples/src/main/java/org/eclipse/rdf4j/examples/TempShaclValidation.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+						return new ValidationResultIterator(iterator, -1);
+					}
+				})
+				.collect(Collectors.toList());
+
+		return new LazyValidationReport(collect, -1);
+
+	}
+
+}
```
Comment (hmottestad) [inline examples/src/main/java/org/eclipse/rdf4j/examples/TempShaclValidation.java:136]: This is the example. It calls into the existing code for parsing shapes and validating data without moving the data around.

It also has limited support for other severity levels.

File: core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/ShaclValidator.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+import org.eclipse.rdf4j.sail.shacl.wrapper.data.ConnectionsGroup;
+import org.eclipse.rdf4j.sail.shacl.wrapper.data.RdfsSubClassOfReasoner;
+import org.eclipse.rdf4j.sail.shacl.wrapper.data.VerySimpleRdfsBackwardsChainingConnection;
+import org.eclipse.rdf4j.sail.shacl.wrapper.shape.CombinedShapeSource;
+import org.eclipse.rdf4j.sail.shacl.wrapper.shape.ShapeSource;
+
+public class ShaclValidator {
+
+	public static ValidationReport validate(SailRepository shapesRepo, SailRepository dataRepo) {
+
```
Comment (hmottestad) [inline core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/ShaclValidator.java:40]: A first attempt at a class that validates two SailRepositories. Should probably use Sail though, since that’s what we end up needing most of the time anyway.

**Comment by hmottestad** [comment]
@abrokenjester I would really like to include this in the 4.3.0 release. I've not quite finished looking through the code myself yet, but I'll make it a priority.

**Comment by hmottestad** [comment]
TODO: Check that shapes stored in the default graph validate the union graph in the data sail.

File: core/model/src/main/java/org/eclipse/rdf4j/model/impl/DynamicModel.java [hunk 3]
```diff
@@ -334,9 +333,9 @@ public Iterable<Statement> getStatements(Resource subject, IRI predicate, Value
 					.createStatement(subject, predicate, object, contexts[0]);
 			Statement foundStatement = statements.get(statement);
 			if (foundStatement == null) {
-				return EmptyIterator::new;
+				return List.of();
 			}
-			return () -> new SingletonIterator<>(foundStatement);
+			return List.of(foundStatement);
```
Comment (hmottestad) [inline core/model/src/main/java/org/eclipse/rdf4j/model/impl/DynamicModel.java:338]: Some unrelated optimisations that I discovered while profiling.

File: core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/ShaclValidator.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+import org.eclipse.rdf4j.sail.shacl.results.lazy.LazyValidationReport;
+import org.eclipse.rdf4j.sail.shacl.results.lazy.ValidationResultIterator;
+import org.eclipse.rdf4j.sail.shacl.wrapper.data.ConnectionsGroup;
+import org.eclipse.rdf4j.sail.shacl.wrapper.data.RdfsSubClassOfReasoner;
+import org.eclipse.rdf4j.sail.shacl.wrapper.data.VerySimpleRdfsBackwardsChainingConnection;
+import org.eclipse.rdf4j.sail.shacl.wrapper.shape.CombinedShapeSource;
+import org.eclipse.rdf4j.sail.shacl.wrapper.shape.ShapeSource;
+
+@Experimental
+public class ShaclValidator {
```
Comment (hmottestad) [inline core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/ShaclValidator.java:39]: This is the new validator class with static methods to allow validation of an existing sail.

File: core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/ast/planNodes/UnionNode.java [hunk 5]
```diff
@@ -58,7 +58,7 @@ public static PlanNode getInstance(PlanNode... nodes) {
 	}
 
 	public static PlanNode getInstanceDedupe(PlanNode... nodes) {
-		PlanNode[] planNodes = Arrays.stream(nodes).filter(n -> !(n instanceof EmptyNode)).distinct().flatMap(n -> {
+		PlanNode[] planNodes = Arrays.stream(nodes).filter(n -> !(n.isGuaranteedEmpty())).distinct().flatMap(n -> {
```
Comment (hmottestad) [inline core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/ast/planNodes/UnionNode.java:61]: Cleaned up some of the code that handles empty interators.

File: core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/results/ValidationResult.java [hunk 6]
```diff
@@ -189,6 +189,13 @@ public final Resource getId() {
 		return id;
 	}
 
+	/**
+	 * @return conforms
+	 */
+	public boolean conforms() {
+		return severity != Severity.Violation;
+	}
```
Comment (hmottestad) [inline core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/results/ValidationResult.java:197]: It never really made sense to support other severity levels before, since anything lower then sh:Violation wouldn't result in a validation exception. Now that we return a validation report regardless I've taken the time to implement support for sh:Warning and sh:Info.

File: core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/wrapper/data/RdfsSubClassOfReasoner.java [hunk 7]
```diff
@@ -149,27 +150,48 @@ private Set<Resource> resolveTypes(Resource value, Map<Resource, Set<Resource>>
...
+	public static RdfsSubClassOfReasoner createReasoner(SailConnection sailConnection,
+			ValidationSettings validationSettings) {
+		return createReasoner(sailConnection, null, validationSettings);
+	}
+
+	public static RdfsSubClassOfReasoner createReasoner(SailConnection sailConnection, SailConnection secondConnection,
+			ValidationSettings validationSettings) {
 		long before = 0;
-		if (shaclSailConnection.sail.isPerformanceLogging()) {
+		if (validationSettings.isPerformanceLogging()) {
```
Comment (hmottestad) [inline core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/wrapper/data/RdfsSubClassOfReasoner.java:161]: One fairly large part of this PR was to eliminate the use of the ShaclSail or ShaclSailConnection in the validation code. So I've introduced some simple config classes to pass in settings that would otherwise be available directly from the ShaclSail or ShaclSailConnection.

File: core/sail/shacl/src/test/java/org/eclipse/rdf4j/sail/shacl/AbstractShaclTest.java [hunk 8]
```diff
@@ -390,6 +405,68 @@ private void printTestCase(TestCase testCase) {
 
 	}
 
+	void runWithShaclValidator(TestCase testCase) {
+
+		SailRepository shapesRepo = new SailRepository(new MemoryStore());
+		SailRepository dataRepo = new SailRepository(new MemoryStore());
+
+		try {
+
```
Comment (hmottestad) [inline core/sail/shacl/src/test/java/org/eclipse/rdf4j/sail/shacl/AbstractShaclTest.java:414]: Integrated the new ShaclValidator code into the existing test suite.

File: core/sail/shacl/src/test/java/org/eclipse/rdf4j/sail/shacl/ShaclValidatorTest.java [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+					Values.iri("http://example.org/g1"));
+		}
+
+		ValidationReport validate = ShaclValidator.validate(data.getSail(), shapes.getSail());
+		assertTrue(validate.conforms());
+	}
+
+	@Test
+	public void testDefaultGraphIsUnionFailureDefaultGraph() throws Exception {
+
```
Comment (hmottestad) [inline core/sail/shacl/src/test/java/org/eclipse/rdf4j/sail/shacl/ShaclValidatorTest.java:64]: Also added some explicit tests for some edge cases with the new ShaclValidator.

File: core/sail/solr/pom.xml [hunk 10]
```diff
@@ -47,18 +47,6 @@
...
-			<groupId>org.slf4j</groupId>
-			<artifactId>log4j-over-slf4j</artifactId>
-			<optional>true</optional>
-		</dependency>
-		<!-- solr-core pulls in log4j2 -->
-		<dependency>
-			<groupId>org.apache.logging.log4j</groupId>
-			<artifactId>log4j-to-slf4j</artifactId>
-			<optional>true</optional>
-		</dependency>
```
Comment (hmottestad) [inline core/sail/solr/pom.xml:61]: These two dependencies were are new and were previously excluded. I'm not sure why they were explicitly added, but I don't see that they are used in any way. Since they are "optional" removing them won't impact our users.

**Comment by hmottestad** [comment]
@abrokenjester I want to create a new milestone build just to test that everything works as intended with GraphDB. So I'm going to merge this PR once the build passes. I'm not trying to bypass code review, it's just that I need to test this with GraphDB before I can start on a 4.3.0 release. If you have any comments or things you want me to change I'll create a followup PR to revert/fix my changes.
